File: c4_linkedlist/L148-m.cpp
//
// Created by LiHeng on 2020/4/19.
//
#include <iostream>
using namespace std;

//  Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

//链表排序-----在 O(n log n) 时间复杂度和常数级空间复杂度下，对链表进行排序。  使用归并排序法
class Solution {
public:
    ListNode* sortList(ListNode* head) {
        if(head == NULL || head->next == NULL )
            return head;
        //利用快慢指针拆分成2个链表
        ListNode* fast = head;
        ListNode* slow = head;
        ListNode* bre = head;
        while(fast != NULL && fast->next !=NULL){
            fast = fast->next->next;
            bre = slow; //断点
            slow = slow->next;
        }

        bre->next = NULL;
        ListNode* l1 = sortList(head);
        ListNode* l2 = sortList(slow);
        return merge2lists(l1, l2);
    }

    ListNode* merge2lists(ListNode* h1, ListNode* h2){
        if(h1==NULL)
            return h2;
        if(h2==NULL)
            return h1;
        ListNode* new_head = new ListNode(0);
        ListNode* temN = new_head;
        while(h1!=NULL&&h2!=NULL){
            if(h1->val > h2->val){
                temN->next = h2;
                h2 = h2->next;
            }
            else{
                temN->next = h1;
                h1 = h1->next;
            }
            temN = temN->next;
        }
        temN->next = (h1 == NULL)? h2:h1;
        ListNode* res = new_head->next;
        delete new_head;
        return res;
    }

};

File: python-algorithm/ds_4_linkedlist/linkedlist_4_others/L148-m.py
#--coding:utf-8--
# 排序链表（归并）
class Solution:
    def sortList(self, head: ListNode) -> ListNode:
        # not head.next 不能没有
        if not head or not head.next: return head
        slow = fast = head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        ano_half = slow.next
        slow.next = None
        lt = self.sortList(head)
        rt = self.sortList(ano_half)
        return self.merge(lt, rt)

    def merge(self, h1, h2):
        if not h1: return h2
        if not h2: return h1
        dummy = ListNode(0)
        cursor = dummy  # 重新开辟一个指针，方便后面返回dummy.next
        while h1 and h2:
            if h1.val <= h2.val:
                cursor.next = h1
                h1 = h1.next
            else:
                cursor.next = h2
                h2 = h2.next
            cursor = cursor.next
        # 太容易漏掉了
        if h1:
            cursor.next = h1
        if h2:
            cursor.next = h2

        return dummy.next
    